src/WarcraftLegacies.Source/Quests/Stormwind/QuestStormwindCity.cs

```diff
@@ -1,62 +1,71 @@
       "ReplaceableTextures\\CommandButtons\\BTNNobbyMansionCastle.blp")
     {
-      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(FourCC("n00V"))));
-      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(FourCC("n00Z"))));
-      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(FourCC("n011"))));
-      AddObjective(new ObjectiveUpgrade(FourCC("h06K"), FourCC("h06K")));
+      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(Constants.UNIT_N00V_DUSKWOOD_10GOLD_MIN)));
+      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(Constants.UNIT_N00Z_ELWYNN_FOREST_20GOLD_MIN)));
+      AddObjective(new ObjectiveControlPoint(ControlPointManager.GetFromUnitType(Constants.UNIT_N011_REDRIDGE_MOUNTAINS_10GOLD_MIN)));
+      AddObjective(new ObjectiveUpgrade(Constants.UNIT_H06K_TOWN_HALL_STORMWIND, Constants.UNIT_H06N_CASTLE_STORMWIND));
       AddObjective(new ObjectiveExpire(1020));
       AddObjective(new ObjectiveSelfExists());
-      foreach (var unit in new GroupWrapper().EnumUnitsInRect(rescueRect).EmptyToList())
-        if (GetOwningPlayer(unit) == Player(PLAYER_NEUTRAL_PASSIVE))
-        {
-          SetUnitInvulnerable(unit, true);
-          _rescueUnits.Add(unit);
-        }
-
-      ResearchId = FourCC("R02S");
+      _rescueUnits = rescueRect.PrepareUnitsForRescue(Player(PLAYER_NEUTRAL_PASSIVE));
+      ResearchId = Constants.UPGRADE_R02S_QUEST_COMPLETED_CLEAR_THE_OUTSKIRTS;
+      Required = true;
     }
 
     //Todo: bad flavour
+    /// <inheritdoc />
     protected override string CompletionPopup =>
       "Stormwind has been liberated, and its military is now free to assist the Alliance.";
 
+    /// <inheritdoc />
     protected override string RewardDescription =>
       "Control of all units in Stormwind and enable Varian to be trained at the altar";
 
+    /// <inheritdoc />
     protected override void OnFail(Faction completingFaction)
     {
-      foreach (var unit in _rescueUnits) unit.Rescue(Player(PLAYER_NEUTRAL_AGGRESSIVE));
+      Player(PLAYER_NEUTRAL_AGGRESSIVE).RescueGroup(_rescueUnits);
     }
 
+    /// <inheritdoc />
     protected override void OnComplete(Faction completingFaction)
     {
-      foreach (var unit in _rescueUnits) unit.Rescue(completingFaction.Player);
-      if (GetLocalPlayer() == completingFaction.Player) PlayThematicMusic("war3mapImported\\StormwindTheme.mp3");
+      if (completingFaction.Player != null)
+        completingFaction.Player.RescueGroup(_rescueUnits);
+
+      if (GetLocalPlayer() == completingFaction.Player) 
+        PlayThematicMusic("war3mapImported\\StormwindTheme.mp3");
     }
 
+    /// <inheritdoc />
     protected override void OnAdd(Faction whichFaction)
     {
-      whichFaction.ModObjectLimit(QuestResearchId, 1);
+      whichFaction.ModObjectLimit(ResearchId, 1);
```

Comment:
You can actually delete this entire method, as the function is already called by virtue of it being set as ResearchId.